bazel/cas/service.go

```diff
@@ -66,27 +98,173 @@ func (s *casServer) GetTree(
+// Writes data into bundlestore from a client via grpc streaming.
 // Implements googleapis bytestream Write
-func (s *casServer) Write(googlebytestream.ByteStream_WriteServer) error {
+// TODO store.Stores do not support partial Writes, and neither does our implementation.
+// We can support partial Write by keeping buffers for inflight requests in the casServer.
+// When the entire Write is buffered, we can Write to the Store and return a response with the result.
+func (s *casServer) Write(ser bytestream.ByteStream_WriteServer) error {
 	log.Info("Received CAS Write request")
-	// Real implementation: stream from client with Recv(), finish with SendAndClose
+
+	if !s.IsInitialized() {
+		return status.Error(codes.Internal, "Server not initialized")
+	}
+
+	var p []byte
+	var buffer *bytes.Buffer
+	var resource *Resource = nil
+	resourceName, storeName := "", ""
+	var err error
+
+	// As indicated above, not supporting partial/resumable Writes for now.
+	// Reads in a stream of data from the client, and proceeds when we've gotten it all.
+	for {
+		wr, err := ser.Recv()
+		if err != nil {
+			return status.Error(codes.Internal, fmt.Sprintf("Failed to Recv: %v", err))
+		}
+
+		// Set up resource on initial WriteRequest. Initialize read buffer based on Digest size
+		if resource == nil {
+			resourceName = wr.GetResourceName()
+
+			resource, err = ParseWriteResource(resourceName)
+			if err != nil {
+				return status.Error(codes.InvalidArgument, fmt.Sprintf("%v", err))
+			}
+			p = make([]byte, 0, resource.Digest.GetSizeBytes())
+			buffer = bytes.NewBuffer(p)
+
+			// If data Exists, terminate immediately with size of existing data (Store is immutable)
+			// Note that Store does not support `stat`, so we trust client-provided size to avoid reading the data
+			storeName = bazel.DigestStoreName(resource.Digest)
+			if exists, err := s.storeConfig.Store.Exists(storeName); err != nil {
+				status.Error(codes.Internal, fmt.Sprintf("Store failed checking existence of %s: %v", storeName, err))
+			} else if exists {
+				log.Infof("Resource exists in store: %s using client digest size: %d", storeName, resource.Digest.GetSizeBytes())
+				res := &bytestream.WriteResponse{CommittedSize: resource.Digest.GetSizeBytes()}
+				err = ser.SendAndClose(res)
+				if err != nil {
+					return status.Error(codes.Internal, fmt.Sprintf("Failed to SendAndClose WriteResponse: %v", err))
+				}
+				return nil
+			}
+		}
+
+		// Validate subsequent WriteRequest fields
+		if wr.GetResourceName() != "" && resourceName != wr.GetResourceName() {
+			return status.Error(codes.InvalidArgument, fmt.Sprintf("ResourceName %s mismatch with previous %s", wr.GetResourceName(), resourceName))
+		}
+		if wr.GetWriteOffset() > 0 {
+			return status.Error(codes.Unimplemented, "Currently unsupported in Scoot - Writes are not resumable")
+		}
+
+		buffer.Write(wr.GetData())
+
+		// Per API, client indicates all data has been sent
+		if wr.GetFinishWrite() {
```

Comment:
As we're not implementing resumable writes, it feels like we probably want to error if this isn't set to true, and always break.

Reply:
I'd like to circle back on resumable writes soon - mainly taking a shortcut there. I think we still need this as even without resumables, we have to loop over the this and check each request that is part of the write stream.

Reply:
I think we shouldn't break when we can reasonably satisfy a request. That said, depending on how long we keep some manner of shortcut, if it should be an error to upload something that's inordinately large or taking too much time between each write.

Reply:
that's not what this is doing, the GetFinishWrite is the end of a normal successful write request. the write server could probably use some timeout or safeguard but this is fine for v1/now